ibtk/src/utilities/HierarchyAveragedDataManager.cpp

```diff
@@ -0,0 +1,384 @@
+// ---------------------------------------------------------------------
+//
+// Copyright (c) 2022 - 2022 by the IBAMR developers
+// All rights reserved.
+//
+// This file is part of IBAMR.
+//
+// IBAMR is free software and is distributed under the 3-clause BSD
+// license. The full text of the license can be found in the file
+// COPYRIGHT at the top level directory of IBAMR.
+//
+// ---------------------------------------------------------------------
+
+/////////////////////////////// INCLUDES /////////////////////////////////////
+#include <ibtk/HierarchyAveragedDataManager.h>
+#include <ibtk/HierarchyGhostCellInterpolation.h>
+#include <ibtk/HierarchyMathOps.h>
+#include <ibtk/snapshot_utilities.h>
+
+#include <HierarchyDataOpsManager.h>
+
+#include <ibtk/namespaces.h> // IWYU pragma: keep
+
+/////////////////////////////// NAMESPACE ////////////////////////////////////
+
+namespace IBTK
+{
+double
+map_to_period(const double t_start, const double t_end, double time)
+{
+    const double period = t_end - t_start;
+    if (IBTK::abs_equal_eps(period, 0.0)) return t_start;
+#ifndef NDEBUG
+    // We only deal with positive period lengths.
+    TBOX_ASSERT(period > 0.0);
+#endif
+
+    while (time > t_end)
+    {
+        // Subtract period until we're less than t_end.
+        time -= period;
+    }
+    while (time < t_start)
+    {
+        // Add period until we're greater than t_start
+        time += period;
+    }
+#ifndef NDEBUG
+    TBOX_ASSERT(time <= t_end && time >= t_start);
+#endif
+    return time;
+}
+
+void
+allocate_patch_data(const int idx,
```

Comment:
We might actually have another version of these two functions floating around somewhere else in the library: to make sure we don't have symbol clashes (now or later) could you put these source-only functions in an anonymous namespace?